File: bindings/java/src/api/com/apple/foundationdb/directory/DirectorySubspace.java
/*
 * DirectorySubspace.java
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.apple.foundationdb.directory;

import static com.apple.foundationdb.directory.DirectoryLayer.EMPTY_BYTES;
import static com.apple.foundationdb.directory.DirectoryLayer.EMPTY_PATH;
import static com.apple.foundationdb.tuple.ByteArrayUtil.printable;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import com.apple.foundationdb.ReadTransactionContext;
import com.apple.foundationdb.TransactionContext;
import com.apple.foundationdb.subspace.Subspace;

/**
 *  A DirectorySubspace represents the <i>contents</i> of a directory, but it
 *  also remembers the path with which it was opened and offers convenience
 *  methods to operate on the directory at that path.
 *
 * <p>
 *     An instance of DirectorySubspace can be used for all the usual subspace
 *     operations. It can also be used to operate on the directory with which
 *     it was opened.
 * </p>
 */
public class DirectorySubspace extends Subspace implements Directory {
	private final List<String> path;
	private final byte[] layer;
	private final DirectoryLayer directoryLayer;

	DirectorySubspace(List<String> path, byte[] prefix, DirectoryLayer directoryLayer) {
		this(path, prefix, directoryLayer, EMPTY_BYTES);
	}

	DirectorySubspace(List<String> path, byte[] prefix, DirectoryLayer directoryLayer, byte[] layer) {
		super(prefix);
		this.path = path;
		this.layer = layer;
		this.directoryLayer = directoryLayer;
	}

	/**
	 * @return a printable representation of this {@code DirectorySubspace}
	 */
	@Override
	public String toString() {
		return getClass().getSimpleName() + '(' + DirectoryUtil.pathStr(path) + ", " + printable(getKey()) + ')';
	}

	/**
	 * Returns whether this {@code DirectorySubspace} is equal to {@code rhs}.
	 * Two {@code DirectorySubspace}s are equal if they were created by the same
	 * {@link DirectoryLayer} and have the same path, layer, and subspace prefix.
	 *
	 * @param rhs the {@code} Object to test for equality
	 * @return true if this is equal to {@code rhs}
	 */
	@Override
	public boolean equals(Object rhs) {
		if(this == rhs) {
			return true;
		}
		if(rhs == null || getClass() != rhs.getClass()) {
			return false;
		}
		DirectorySubspace other = (DirectorySubspace)rhs;
		return (path == other.path || path.equals(other.path)) &&
			   Arrays.equals(layer, other.layer) &&
			   directoryLayer.equals(other.directoryLayer) &&
			   super.equals(rhs);
	}

	/**
	 * Computes a hash code compatible with the {@link #equals(Object) equals()} method of
	 * this class. In particular, it will produce a hash code that is based off of the hashes
	 * of its path, its layer, and its subspace prefix.
	 *
	 * @return a hash compatible with this class's {@code equals()} method
	 */
	@Override
	public int hashCode() {
		return path.hashCode() ^ (Arrays.hashCode(layer) * 1153) ^ (directoryLayer.hashCode() * 929) ^ (super.hashCode() * 419);
	}

	@Override
	public List<String> getPath() {
		return Collections.unmodifiableList(path);
	}

	@Override
	public byte[] getLayer() {
		return Arrays.copyOf(layer, layer.length);
	}

	@Override
	public DirectoryLayer getDirectoryLayer() {
		return directoryLayer;
	}

	@Override
	public CompletableFuture<DirectorySubspace> createOrOpen(TransactionContext tcx, List<String> subpath, byte[] otherLayer) {
		return directoryLayer.createOrOpen(tcx, getPartitionSubpath(subpath), otherLayer);
	}

	@Override
	public CompletableFuture<DirectorySubspace> open(ReadTransactionContext tcx, List<String> subpath, byte[] otherLayer) {
		return directoryLayer.open(tcx, getPartitionSubpath(subpath), otherLayer);
	}

	@Override
	public CompletableFuture<DirectorySubspace> create(TransactionContext tcx, List<String> subpath, byte[] otherLayer, byte[] prefix) {
		return directoryLayer.create(tcx, getPartitionSubpath(subpath), otherLayer, prefix);
	}

	@Override
	public CompletableFuture<List<String>> list(ReadTransactionContext tcx, List<String> subpath) {
		return directoryLayer.list(tcx, getPartitionSubpath(subpath));
	}

	@Override
	public CompletableFuture<DirectorySubspace> move(TransactionContext tcx, List<String> oldSubpath, List<String> newSubpath) {
		return directoryLayer.move(tcx, getPartitionSubpath(oldSubpath), getPartitionSubpath(newSubpath));
	}

	@Override
	public CompletableFuture<DirectorySubspace> moveTo(TransactionContext tcx, List<String> newAbsolutePath) {
		DirectoryLayer dir = getLayerForPath(EMPTY_PATH);
		int partitionLen = dir.getPath().size();
		List<String> partitionPath = newAbsolutePath.subList(0, Math.min(newAbsolutePath.size(), partitionLen));
		if(!partitionPath.equals(dir.getPath()))
			throw new DirectoryMoveException("Cannot move between partitions", path, newAbsolutePath);

		return dir.move(tcx,
						getPartitionSubpath(EMPTY_PATH, dir),
						newAbsolutePath.subList(partitionLen, newAbsolutePath.size()));
	}

	@Override
	public CompletableFuture<Void> remove(TransactionContext tcx, List<String> subpath) {
		DirectoryLayer dir = getLayerForPath(subpath);
		return dir.remove(tcx, getPartitionSubpath(subpath, dir));
	}

	@Override
	public CompletableFuture<Boolean> removeIfExists(TransactionContext tcx, List<String> subpath) {
		DirectoryLayer dir = getLayerForPath(subpath);
		return dir.removeIfExists(tcx, getPartitionSubpath(subpath, dir));
	}

	@Override
	public CompletableFuture<Boolean> exists(ReadTransactionContext tcx, List<String> subpath) {
		DirectoryLayer dir = getLayerForPath(subpath);
		return dir.exists(tcx, getPartitionSubpath(subpath, dir));
	}

	private List<String> getPartitionSubpath(List<String> path) {
		return getPartitionSubpath(path, directoryLayer);
	}

	private List<String> getPartitionSubpath(List<String> path, DirectoryLayer directoryLayer) {
		return PathUtil.join(this.path.subList(directoryLayer.getPath().size(), this.path.size()), path);
	}

	/**
	 * Called by all functions that could operate on this subspace directly (moveTo, remove, removeIfExists, exists).
	 * Subclasses can chooose to return a different directory layer to use for the operation if path is in fact empty.
	 */
	DirectoryLayer getLayerForPath(List<String> path) {
		return directoryLayer;
	}
}

File: bindings/go/src/fdb/directory/directory_subspace.go
/*
 * directory_subspace.go
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2024 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// FoundationDB Go Directory Layer

package directory

import (
	"context"
	"fmt"
	"strings"

	"github.com/apple/foundationdb/bindings/go/src/fdb"
	"github.com/apple/foundationdb/bindings/go/src/fdb/subspace"
)

// DirectorySubspace represents a Directory that may also be used as a Subspace
// to store key/value pairs. Subdirectories of a root directory (as returned by
// Root or NewDirectoryLayer) are DirectorySubspaces, and provide all methods of
// the Directory and subspace.Subspace interfaces.
type DirectorySubspace interface {
	subspace.Subspace
	Directory
}

type directorySubspace struct {
	subspace.Subspace
	dl    directoryLayer
	path  []string
	layer []byte
}

// String implements the fmt.Stringer interface and returns human-readable
// string representation of this object.
func (ds directorySubspace) String() string {
	var path string
	if len(ds.path) > 0 {
		path = "(" + strings.Join(ds.path, ",") + ")"
	} else {
		path = "nil"
	}
	return fmt.Sprintf("DirectorySubspace(%s, %s)", path, fdb.Printable(ds.Bytes()))
}

func (d directorySubspace) CreateOrOpen(ctx context.Context, t fdb.Transactor, path []string, layer []byte) (DirectorySubspace, error) {
	return d.dl.CreateOrOpen(ctx, t, d.dl.partitionSubpath(d.path, path), layer)
}

func (d directorySubspace) Create(ctx context.Context, t fdb.Transactor, path []string, layer []byte) (DirectorySubspace, error) {
	return d.dl.Create(ctx, t, d.dl.partitionSubpath(d.path, path), layer)
}

func (d directorySubspace) CreatePrefix(ctx context.Context, t fdb.Transactor, path []string, layer []byte, prefix []byte) (DirectorySubspace, error) {
	return d.dl.CreatePrefix(ctx, t, d.dl.partitionSubpath(d.path, path), layer, prefix)
}

func (d directorySubspace) Open(ctx context.Context, rt fdb.ReadTransactor, path []string, layer []byte) (DirectorySubspace, error) {
	return d.dl.Open(ctx, rt, d.dl.partitionSubpath(d.path, path), layer)
}

func (d directorySubspace) MoveTo(ctx context.Context, t fdb.Transactor, newAbsolutePath []string) (DirectorySubspace, error) {
	return moveTo(ctx, t, d.dl, d.path, newAbsolutePath)
}

func (d directorySubspace) Move(ctx context.Context, t fdb.Transactor, oldPath []string, newPath []string) (DirectorySubspace, error) {
	return d.dl.Move(ctx, t, d.dl.partitionSubpath(d.path, oldPath), d.dl.partitionSubpath(d.path, newPath))
}

func (d directorySubspace) Remove(ctx context.Context, t fdb.Transactor, path []string) (bool, error) {
	return d.dl.Remove(ctx, t, d.dl.partitionSubpath(d.path, path))
}

func (d directorySubspace) Exists(ctx context.Context, rt fdb.ReadTransactor, path []string) (bool, error) {
	return d.dl.Exists(ctx, rt, d.dl.partitionSubpath(d.path, path))
}

func (d directorySubspace) List(ctx context.Context, rt fdb.ReadTransactor, path []string) ([]string, error) {
	return d.dl.List(ctx, rt, d.dl.partitionSubpath(d.path, path))
}

func (d directorySubspace) GetLayer() []byte {
	return d.layer
}

func (d directorySubspace) GetPath() []string {
	return d.path
}
